week 4 & 5

bishoy-magdy opened the pull request:
# What has been done
- Rename `emailParser` file to `emailConfigure`.
- Installed IMAP and its requirements and made the first IMAP connection in the plugin (SSL/TLS).
- Added the manual connection screen.
- Created the Main Screen.
- Open/Close Monitoring based on 'from'.
- Logout button
- Created an error alarm, a warning alarm when an error is caused while logging monitoring (using alert)
***
Make sure when logging in to Gmail using the app password and turning on imap, but other email providers, just create the app password.

**To turn on imap** 
<details>

 1. On your computer, open Gmail.

2. In the top right, click Settings Settings and then See all  settings.

3. Click the Forwarding and POP/IMAP tab.

4. In the "IMAP access" section, select Enable IMAP.

5. Click Save Changes.
</details>

**To create an app password [link](https://support.google.com/mail/answer/185833?hl=en#:~:text=Create%20%26%20use%20App%20Passwords)**

bishoy-magdy committed:
update - week4&5

roman-r-m commented on src/core/imap.ts:
```diff
@@ -0,0 +1,130 @@
+                ({ mailBox, criteria } = this.query);
+
+                // Check if the connection is still stable.
+                await this.state().then(async () => {
```
isn't this the same as
```
await this.state()
await this.openBox(mailBox)
let messages = await this.search(criteria)
...
```

bishoy-magdy replied:
> isn't this the same as

Yes, it's the same thing.

> let messages = await this.search(criteria)

But if this method is used, a new variable will be created in each monitoring process.

However, there is another way to declare this variable above the monitoring and reassignment it every time.
Do I change it this way?

roman-r-m replied:
> But if this method is used, a new variable will be created in each monitoring process.

Is it a problem?

bishoy-magdy replied:
> Is it a problem?

I am not sure 100%, but I think it may cause memory leaks until the user logs out.
But I can declare the variable before monitoring to avoid this problem!

roman-r-m replied:
A variable on its own does not use memory. It's the data that is returned from `this.search`. Declaring a variable in advance changes nothing.

bishoy-magdy replied:
> A variable on its own does not use memory. It's the data that is returned from `this.search`. Declaring a variable in advance changes nothing.

OK, I get it now .
I have replaced 'then' with 'await'.

bishoy-magdy committed:
update - week4&5

laurent22 commented:
@bishoy-magdy, could you setup some linter on your repo as indentation is inconsistent? At a minimum it should use tabs and I feel things like strict equality should be required. See if you can use the eslintc file from the main repo: https://github.com/laurent22/joplin/blob/dev/.eslintrc.js If it's too complicated to get it working you can disable some of the rules.

bishoy-magdy committed:
add eslint

bishoy-magdy commented:
> @bishoy-magdy, could you setup some linter on your repo as indentation is inconsistent? At a minimum it should use tabs and I feel things like strict equality should be required. See if you can use the eslintc file from the main repo: https://github.com/laurent22/joplin/blob/dev/.eslintrc.js If it's too complicated to get it working you can disable some of the rules.

Yes, sure. I added eslint with Google extend and added some rules from the Joplin repo.

If there is something else, please tell me. Thanks.

roman-r-m commented on src/core/imap.ts:
```diff
@@ -0,0 +1,117 @@
+            } else {
+                try {
+                    await this.init();
+                    resolve('Reconnected successfully');
```
You resolve this passing string but never use the return value.

roman-r-m replied:
I think this is where having type annotations could help, then it'd be immediately clear that you don't use the return type.

bishoy-magdy replied:
> You resolve this passing string but never use the return value.

I was trying to make the function more readable without comments.

bishoy-magdy replied:
> I think this is where having type annotations could help, then it'd be immediately clear that you don't use the return type.

Yes, I annotated the promises by void type.

roman-r-m commented on src/core/imap.ts:
```diff
@@ -0,0 +1,117 @@
+                    // Check if the connection is still stable.
+                    await this.state();
+
+                    await this.openBox(mailBox);
```
Same here, this returns `box` that you never use.

bishoy-magdy replied:
Done :)

PackElend replied:
how does the change look like I cannot find anything in the latest commit? 
It could be that I am just too much of coding novice to see it :)

bishoy-magdy replied:
> how does the change look like I cannot find anything in the latest commit? It could be that I am just too much of coding novice to see it :)

Briefly, Dealing with operations in imap is asynchronous because we are waiting for a response from the mail service provider with approval or rejection, so this is not an instantaneous process.

That is why you will notice the word 'await' beside the function.

When this data is available from the email service provider, we use the 'resolve' and can also return the box or message from the email provider depending on what the function does.

But we don't need this data now, so what I [did](https://github.com/joplin/plugin-email/pull/4/commits/e8784fec62dec940231c443f06a80289ae8b5780) was just `resolve()` without value.

```
Before:
	resolve(box);
After:
	resolve();
```
If there is anything unclear, please tell me.

PackElend replied:
thx I found the change in the code, eventually, in https://github.com/joplin/plugin-email/pull/4/commits/e8784fec62dec940231c443f06a80289ae8b5780, thx for pushing me in the right direction

thx for the nice explanation :), can you show me where resolve() is defined?

bishoy-magdy replied:
You’re welcome,

> can you show me where resolve() 

###  Inside promise function 
```
let promise = new Promise(function(`resolve`, reject) {

});
```
Resolve and reject are callback functions, and calling the resolve function if the job is finished successfully.

If you are interested in the promise, this [website](https://javascript.info/promise-basics) is very nice.

roman-r-m commented:
One thing I've noticed when trying the plugin is that the panel opens on every start even if I closed it before. Ideally I think the plugin should remember this and not reopen the panel

bishoy-magdy committed:
update - imap Promises

bishoy-magdy commented:
> One thing I've noticed when trying the plugin is that the panel opens on every start even if I closed it before. Ideally I think the plugin should remember this and not reopen the panel

Yes, also this problem occurs when the main screen starts and enters the email and closes the panel, which when opened again will open the main screen without saving the email.

I mentioned this problem in the weekly report and I will solve this problem this week.

bishoy-magdy commented:
> One thing I've noticed when trying the plugin is that the panel opens on every start even if I closed it before. Ideally I think the plugin should remember this and not reopen the panel

I am trying to produce this issue on Linux and Windows, but the plugin remembers its last state and work fine, so is there a way to produce this issue?

https://user-images.githubusercontent.com/58605547/182437506-0db6b3d2-3cbe-4990-80d4-6396a032d90d.mp4

roman-r-m commented:
Yes, you're right. Not sure what I did back then but now it seems to be working correctly.

I also see this hasn't been merged - will do now.
